### ISSUE TITLE
Possible Feature: Indicator Tootip

### ISSUE THREAD

**Issue by JasonLG1979** [JasonLG1979/gnome-shell-extension-mpris-indicator-button#16]
A lot of users expressed interest in having track info in the panel, and it get it, it's nice to be able to see it at a glance. But I did not implement that feature because of how it would possibly affect other indicators in the panel.

As a compromise I thought that maybe users would like a tool tip that showed that info on mouse over of the indicator?

Something like this:
![Screenshot from 2019-10-16 18-00-16](https://user-images.githubusercontent.com/6667703/66966109-b8792b80-f041-11e9-8094-d0c5d7e846d7.png)
![Screenshot from 2019-10-16 17-59-30](https://user-images.githubusercontent.com/6667703/66966112-badb8580-f041-11e9-8c4d-304a5bca478d.png)

GNOME Shell has no built in tool tip functionality that I know of, so before I finish writing it from scratch I thought I'd get your thoughts.

**Comment by JasonLG1979** [comment]
Well I've managed to hack together working version complete with nice fade in and out animations. It even works reasonably well with the dash to dock extension, even when the indicator is at the bottom of the screen.

@terencode you were very helpful in polishing up the last release. What are your thoughts on this? Good? Bad? Indifferent?

**Comment by JasonLG1979** [comment]
This is what I have. Like I said there is no built in tool tip functionality so basically I just poop a label on the screen and then track the state size and position of the indicator and do a little math and bingo a tool tip that *appears* to be attached to the indicator.

```js
const ToolTip = GObject.registerClass({
    GTypeName: "ToolTip"
}, class TrackInfo extends St.Label {
    _init(indicator) {
        super._init({
            style_class: "osd-window tool-tip",
            visible: false
        });

        this._indicator = indicator;

        this._indicatorHandlerIds = [
            this._indicator.connect("current-artist-title", (indicator, artist, title) => {
                this.text = title ? `${artist} • ${title}` : `${artist}`;
                if (this.visible) {
                    this._updatePosition();
                }
            }),

            this._indicator.connect("notify::hover", () => {
                if (this._indicator.hover && !this._indicator.menu.isOpen && !this.visible) {
                    this._show();
                    this._updatePosition();
                } else if (this.visible) {
                    this._hide();
                }
            }),

            this._indicator.connect("notify::visible", () => {
                if (!this._indicator.visible && this.visible) {
                    this._hide();
                }
            })
        ];

        this._indicatorMenuHandlerIds = [
            this._indicator.menu._onMenuOpenStateChangedId = indicator.menu.connect('open-state-changed', () => {
                if (this.visible) {
                    this._hide();
                }
            }),

            this._indicator.menu.connect("destroy", () => {
                this._indicatorHandlerIds.forEach(id => this._indicator.disconnect(id));
                this._indicatorMenuHandlerIds.forEach(id => this._indicator.menu.disconnect(id));
                this._indicatorMenuHandlerIds = null;
                this._indicatorHandlerIds = null;
                this._indicator = null;
                Main.uiGroup.remove_actor(this);
            })
        ];

        Main.uiGroup.add_actor(this);
    }

    _updatePosition() {
        let monitor = Main.layoutManager.primaryMonitor;
        let scaleFactor = St.ThemeContext.get_for_stage(global.stage).scale_factor;
        let [, , indWidth, indHeight] = this._indicator.get_preferred_size();
        let m = Math.round((Math.min(indWidth, indHeight) * 2) / scaleFactor);
        this.style = `margin: ${m}px;`;
        let [X, Y] = this._indicator.get_transformed_position();
        let [, , thisWidth, thisHeight] = this.get_preferred_size();
        X = Math.round(Math.max(Math.min(X + ((indWidth - thisWidth) / 2), monitor.width - thisWidth), 0));
        Y = Math.round(Math.max(Math.min(Y + ((indHeight - thisHeight) / 2), monitor.height - thisHeight), 0));
        this.set_position(X, Y); 
    }

    _show() {
        this.show();
        this.opacity = 0;
        this.ease({
            opacity: 255,
            duration: DASH_ITEM_LABEL_SHOW_TIME,
            mode: Clutter.AnimationMode.EASE_OUT_QUAD
        });
    }

    _hide() {
        this.ease({
            opacity: 0,
            duration: DASH_ITEM_LABEL_HIDE_TIME,
            mode: Clutter.AnimationMode.EASE_OUT_QUAD,
            onComplete: () => this.hide()
        });
    }
});
```

**Comment by terencode** [comment]
Hey it's really not bad! Here is my feedback:

- Show if the music is paused: `Artist Name · Music Name (paused)`
- When skipping tracks, the tooltip needs to adjust to the title length but you can see it moving to the left/right. Would it be possible to only display it once the calculation is made?

**Comment by JasonLG1979** [comment]
> * Show if the music is paused: `Artist Name · Music Name (paused)`

I thought about that but it seemed redundant. If the music is paused then well you'll know it's paused because you won't hear anything.

> * When skipping tracks, the tooltip needs to adjust to the title length but you can see it moving to the left/right. Would it be possible to only display it once the calculation is made?

It tries to center itself to the indicator if there is enough room. The margins are based on the size of the indicator (the box not the icon) plus 0.3em. So for example if the indicator is 16 x 32 then the margins would be 20 x 36.  

Here is the line that determines horizontal placement. 

```js                                                     
x = Math.round(Math.max(Math.min(x + ((indWidth - thisWidth) / 2), monitor.width - indWidth - thisWidth), indWidth));

```

**Comment by terencode** [comment]
> I thought about that but it seemed redundant. If the music is paused then well you'll know it's paused because you won't hear anything.

Not if you aren't using your headphones.

For the placement, it's fine, it's just you see it wrongly placed in the first place and then it corrects itself and it would be better if you would see it in the correct position from the beginning.

**Comment by JasonLG1979** [comment]
> Not if you are using your headphones and not wearing them ^^ .

Very true. I'll see what I can come up with.

> For the placement, it's fine, it's just you see it wrongly placed in the first place and then it corrects itself and it would be better if you would see it in the correct position from the beginning.

What do you mean? As you change tracks with it shown the tooltip will move if the text changes size. As I said it will always try to be centered on the indicator. But if there is not enough room the tooltip will bump up against the margin.


For example there's enough room to center the tooltip:
![Screenshot from 2019-10-16 18-00-16](https://user-images.githubusercontent.com/6667703/66966109-b8792b80-f041-11e9-8094-d0c5d7e846d7.png)

Not enough room to center the tool tip so the right side is bumped against the margin but the tooltip is still under the indicator.
![Screenshot from 2019-10-16 17-59-30](https://user-images.githubusercontent.com/6667703/66966112-badb8580-f041-11e9-8c4d-304a5bca478d.png)

**Comment by JasonLG1979** [comment]
As far as playback status it will more than likely be an icon representing the state, Playing, Paused or Stopped.

So it would go:

 *icon* <*Artist*> • <*Title*>

**Comment by terencode** [comment]
> As far as playback status it will more than likely be an icon representing the state, Playing, Paused or Stopped.
> 
> So it would go:
> 
> _icon_ <_Artist_> • <_Title_>

Makes sense to me.

When trying to reproduce the tooltip issue, I discovered that doing fast previous/next using the keybindings produces the following:
```
oct. 20 00:42:33 terence-arch gnome-shell[1814]: JS ERROR: Extension mprisindicatorbutton@JasonLG1979.github.io: Gio.IOErrorEnum: Operation not supported
                                                 getCover/<@/usr/share/gnome-shell/extensions/mprisindicatorbutton@JasonLG1979.github.io/widgets.js:100:45
oct. 20 00:42:33 terence-arch gnome-shell[1814]: JS ERROR: Extension mprisindicatorbutton@JasonLG1979.github.io: Gio.IOErrorEnum: Operation not supported
                                                 getCover/<@/usr/share/gnome-shell/extensions/mprisindicatorbutton@JasonLG1979.github.io/widgets.js:100:45
oct. 20 00:42:33 terence-arch gnome-shell[1814]: JS ERROR: Extension mprisindicatorbutton@JasonLG1979.github.io: Gio.IOErrorEnum: Error opening file /home/terence/.cache/MellowPlayer/Covers/78661033: No such file or directory
                                                 getCover/<@/usr/share/gnome-shell/extensions/mprisindicatorbutton@JasonLG1979.github.io/widgets.js:100:45
```

The video is coming.

**Comment by JasonLG1979** [comment]
> When trying to reproduce the tooltip issue, I discovered that doing fast previous/next using the keybindings produces the following:
> 
> ```
> oct. 20 00:42:33 terence-arch gnome-shell[1814]: JS ERROR: Extension mprisindicatorbutton@JasonLG1979.github.io: Gio.IOErrorEnum: Operation not supported
>                                                  getCover/<@/usr/share/gnome-shell/extensions/mprisindicatorbutton@JasonLG1979.github.io/widgets.js:100:45
> oct. 20 00:42:33 terence-arch gnome-shell[1814]: JS ERROR: Extension mprisindicatorbutton@JasonLG1979.github.io: Gio.IOErrorEnum: Operation not supported
>                                                  getCover/<@/usr/share/gnome-shell/extensions/mprisindicatorbutton@JasonLG1979.github.io/widgets.js:100:45
> oct. 20 00:42:33 terence-arch gnome-shell[1814]: JS ERROR: Extension mprisindicatorbutton@JasonLG1979.github.io: Gio.IOErrorEnum: Error opening file /home/terence/.cache/MellowPlayer/Covers/78661033: No such file or directory
>                                                  getCover/<@/usr/share/gnome-shell/extensions/mprisindicatorbutton@JasonLG1979.github.io/widgets.js:100:45
> ```
> 
> The video is coming.

Those are IO error when it comes to fetching cover art. They're not fatal errors, more of a warning really. They're handled correctly in the extension. Other then the log messages and the extension using a fallback icon for the cover it's not going to affect anything else.

**Comment by JasonLG1979** [comment]
When I say handled correctly I mean that the errors won't propagate or cause other errors. It's not going to affect the rest of the Shell at all.

**Comment by JasonLG1979** [comment]
Oh and I went  ahead and added that playback status icon you wanted.

**Comment by JasonLG1979** [comment]
the code that handles the potential error is:
```js
                    try {
                        let [bytes, etag] = file.load_bytes_finish(result);
                        let fingerPrint = cover_url + etag;
                        this._fireCallbacks(cover_url, bytes, fingerPrint);
                    } catch (error) {
                        if (!error.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.CANCELLED)) {
                            this._fireCallbacks(cover_url, null, null);
                            logMyError(error);
                        }
                    }
```

the logging function is:
```js
function logMyError(error) {
    // Cancelling counts as an error don't spam the logs.
    if (!error.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.CANCELLED)) {
        logError(error, `Extension ${UUID}`);
    }
}
```

**Comment by terencode** [comment]
Thanks. 
Yes, I know it won't affect anything, but it will make me and my log  happy you check if the file exists before trying to use it :3

**Comment by JasonLG1979** [comment]
> Yes, I know it won't affect anything, but it will make me and my log happy you check if the file exists before trying to use it :3

Checking to see if the file exists is a no no because:

1. Checking to see if it exists is blocking IO. Blocking IO is the devil. Since the extension runs in the same thread as the Shell it would freeze the Shell until complete.

2. Checking to see if a file exists before you try to open it is racey. Let's say you check and it's there but in between your check it disappears now you code that assumes that it's there because you checked fails. It's better to beg forgiveness than ask permission as they say.

3. That's not my job. If the file doesn't exist then the player should not have sent the url in the metadata.

**Comment by JasonLG1979** [comment]
For more on checking to see if a file exists:

https://lazka.github.io/pgi-docs/#Gio-2.0/classes/File.html#Gio.File.query_exists

It's the python docs but that doesn't matter really the idea is the same.

**Comment by terencode** [comment]
Ah ok I understand better what you mean, thanks. (to not block IO, can't you use async functions like  GLib.spawn_command_line_async ) ?

By the way, it seems like the indicators are mismatched ( the triangle instead of || and vice-versa)

**Comment by JasonLG1979** [comment]
> Ah ok I understand better what you mean, thanks. (to not block IO, can't you use async functions like GLib.spawn_command_line_async ) ?

I don't need to the act of trying to read the file will tell me if it exists.

> By the way, it seems like the indicators are mismatched ( the triangle instead of || and vice-versa)

It would not match the play/pause button. The tooltip icon shows what is as in the current playback status. A play icon if it's playing and a paused icon if it's paused. The icon on the button shows what it does. So if the player is playing there's a pause icon in the play/pause icon because that's what it does when you click it.

**Comment by JasonLG1979** [comment]
I log that specific bit on failure because aside from the cover art all the other props are basically simple types like strings and doubles and whatnot that can be used as-is. Cover art is the most complex (and thus potentially error prone) part of mpris. It's sent as a url string but it's up to the client to get the image.

The above error message is valid and log worthy. The player sent a url to an image file that doesn't exist. If I had to guess, if the file ever existed at all that the player may have deleted the cached image before it sent another metadata update with a different image url. Which would be a race condition bug on the part of the player.

**Comment by terencode** [comment]
> It would not match the play/pause button.

Ah yeah makes sense.

> I don't need to the act of trying to read the file will tell me if it exists.

I know, this makes sense as well, but I was wondering in an unrelated manner if using `GLib.spawn_command_line_async`  would allow you to not block the IO in a different context.

And your last comment also makes sense ^^ I'll report the issue to the corresponding player.

**Comment by JasonLG1979** [comment]
> I know, this makes sense as well, but I was wondering in an unrelated manner if using `GLib.spawn_command_line_async` would allow you to not block the IO in a different context.

Checking to see if the file exists before trying to read the file, which also checks to see if the file exists is redundant. And doubles IO time. Which is a big deal if the player send a remote cover url.

> And your last comment also makes sense ^^ I'll report the issue to the corresponding player.

They just need to use common sense. Before metadata is sent with a cover url you better damn well make sure you're done writing the file. And before you delete the file send metadata that at least invalidates the url.

**Comment by terencode** [comment]
> Checking to see if the file exists before trying to read the file, which also checks to see if the file exists is redundant. And doubles IO time. Which is a big deal if the player send a remote cover url.

Arghhh you don't get me. I'm not talking about checking if a file exists, I'm talking for any other legit operation that could block IO.

**Comment by JasonLG1979** [comment]
I don't do blocking IO. I use all async functions.

**Comment by JasonLG1979** [comment]
I even wrote a special class that handles getting the covers that stops redundant async IO.

**Comment by terencode** [comment]
Ok cool, so there is a way to not block IO, that was my question.

**Comment by JasonLG1979** [comment]
> Ok cool, so there is a way to not block IO, that was my question.

Yes. Pretty much all IO related functions in Gio have both blocking (synchronous) and non-blocking (asynchronous) versions.

**Comment by JasonLG1979** [comment]
Basically in the background it does the IO in another thread. When it gets the result it fires a callback.

**Comment by JasonLG1979** [comment]
@terencode test what I have right now if you would please and let me know what you think and how everything looks. If all is well I'm going to go ahead and close this and make it as implemented/complete.

**Comment by terencode** [comment]
Working pretty darn good :+1:

**Comment by JasonLG1979** [comment]
> Working pretty darn good

I tested against the latest git master of dash to dock to make sure it works with panels on any side of the screen. Who'd a thought a tooltip could be so complicated,lol!!!

**Comment by terencode** [comment]
Ahah well done! Maybe open an issue for gnome-shell to support tool tips natively?

**Comment by JasonLG1979** [comment]
> Ahah well done! Maybe open an issue for gnome-shell to support tool tips natively?

I have a feeling it's a design choice not to.

**Comment by JasonLG1979** [comment]
![Screenshot from 2019-10-21 20-13-22](https://user-images.githubusercontent.com/6667703/67253976-278ac180-f440-11e9-9b2b-0de757576878.png)
![Screenshot from 2019-10-21 20-18-01](https://user-images.githubusercontent.com/6667703/67253977-278ac180-f440-11e9-88cd-de0cbe93b31d.png)
![Screenshot from 2019-10-21 20-18-19](https://user-images.githubusercontent.com/6667703/67253978-278ac180-f440-11e9-8663-61ef6d9a87e2.png)
![Screenshot from 2019-10-21 20-18-34](https://user-images.githubusercontent.com/6667703/67253979-28235800-f440-11e9-8da3-b0b91b9b1451.png)

**Comment by JasonLG1979** [comment]
There you go @terencode I added a fly in and out animations. The pivot point of the animation is set to whatever side the panel is on so as to appear as though it's kinda coming from the indicator. The animation happens at show, hide and song and playback status changes.

**Comment by JasonLG1979** [comment]
It could be as easy as:
```js
let toolTip = new ToolTip(indicator, "Whatever Text", "icon-name");

toolTip.update("Different Text", "different-icon-name");
```

**Comment by terencode** [comment]
Wow really great work! 
Could this be integrated to a gnome-shell library? or least maybe create a separate repo ?
Testing right now.

**Comment by JasonLG1979** [comment]
>Could this be integrated to a gnome-shell library?

Once upon a time there was a native tooltip implementation:
 https://gitlab.gnome.org/GNOME/gnome-shell/commit/d81958a07498f3076f19e95847345a24e18bdfdd

>least maybe create a separate repo ?

I already though about that. I'll need to make it more configurable form the outside. Right now implementation is internally configured, because it's written for my particular use and taste. Not everyone would want an icon or animation for example. There's also a matter of styling. In my case I used the osd window as a base style and overrode a bunch of it.

**Comment by JasonLG1979** [comment]
I really would like to test it on a multi monitor set up. In theory and by the book it should work but I don't have a way to test it as I do not have a multi monitor set up.

**Comment by terencode** [comment]
Nor do I have one unfortunately. Do as you think is best. I'll see if GNOME can consider adding them back.

**Comment by JasonLG1979** [comment]
> I'll see if GNOME can consider adding them back.

I doubt they will. This implementation is only useful for panel items, indicator buttons in particular.

**Comment by JasonLG1979** [comment]
In a static or at least known layout as the Shell is for the Shell devs it's not particularly needed. They can assume an awful lot. They only thing that my implementation assumes is that the indicator will be on the edge of the monitor.

**Comment by JasonLG1979** [comment]
There @terencode I made an abstract ```ToolTipBase``` class that in combination with the ```ToolTipConstraint``` class (which you would never have to touch) makes it pretty much copy and paste for someone else to use it if they like. My concrete ```ToolTip``` is only 50 lines of code.

If you look at the code I made it stupid easy. <b>TONS</b> of convenience functions and it's self cleaning. There's only 1 mandatory override. (the hover notify handler).

**Comment by terencode** [comment]
Very cool! I already have an idea for an other extension where this could be implemented!

**Comment by JasonLG1979** [comment]
> Very cool! I already have an idea for an other extension where this could be implemented!

Right on. Which one?

And btw ofc It's GPL3 so anyone can use it. And I'm not married to it, so If someone wants to change it in anyway or use it just in part it's not going to hurt my feelings. Basically it's just out there and if anyone can make use of it cool. And also if someone finds a better way I'm open to that too.

**Comment by terencode** [comment]
https://github.com/andyholmes/gnome-shell-extension-gsconnect which would show device name and battery remaining when hovering the device icon.

That's really appreciated :)

**Comment by JasonLG1979** [comment]
> https://github.com/andyholmes/gnome-shell-extension-gsconnect which would show device name and battery remaining when hovering the device icon.

Well put in a feature request. I'd love to get Andy's feedback. He seems like he knows his way around the GNOME Shell code and a thing or 2 about writing extensions. In fact he's the guy that approves at least my extensions on the extensions site.

**Comment by andyholmes** [comment]
(prelude: sorry, I wrote a short response days ago but didn't click comment, it was just sitting in the tab!)

That looks great! I like the use of ClutterContraints, Abstract classes and especially the Atk roles (I think support for abstract's is pretty new though?). I'd actually like to try this when I get time to see if a screen reader can hack it.

I have a much less modern [tooltip class](https://github.com/andyholmes/gnome-shell-extension-gsconnect/blob/master/src/shell/tooltip.js), but I was keep a little closer to GtkTooltip's API. One neat thing they do is browse mode, which means hovering between 2+ widgets with tooltips brings the time-in to 60ms, so once *a* tooltip is open you don't have to wait each time.

> Something is broken in Gio in regards to using custom/local icons (code comment)

I think this is related to [!317](https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/317); I believe GNOME Shell doesn't use the "global/session" GtkIconTheme object so doesn't get custom icons in some cases. I think I asked Carlos Garnacho about making GS's GtkIconTheme public and he seemed open to it, I just never followed up. IIRC, this was another "remove X-specific code" change.

**Comment by JasonLG1979** [comment]
> That looks great! I like the use of ClutterContraints, Abstract classes and especially the Atk roles

Thanks, that means a lot. It was a lot of head scratching because I'm not what you'd call a pro by any stretch of the word and the docs are few and far between. I mostly read the [python doc]( https://lazka.github.io/pgi-docs/) and look at how things are used in the gnome-shell source.

> I'd actually like to try this when I get time to see if a screen reader can hack it.

I tested with the screen reader and by default no tooltips are read aloud even if I hookup the Atk objects with Atk.Object.set_parent and Atk.Object.add_relationship. I looked at the gtk source code and all they do is set the description of the target to whatever is in the tooltip. But it really doesn't matter because tooltips in GTK apps aren't read aloud either unless you check an extra box in the orca UI (I don't remember which one sorry) But once that box is checked hooking up the Atk objects is unnecessary because the screen reader will read whatever pops up on screen.

> One neat thing they do is browse mode, which means hovering between 2+ widgets with tooltips brings the time-in to 60ms, so once _a_ tooltip is open you don't have to wait each time.

The tooltip is not really a child of the indicator but it uses the indicator to decide where to put itself and what to do when the indicator goes away or when/if the indicator's menu opens. I guess it could be shared among multiple indicators by resetting the indicator something like:

```js

set_source(indicator) {
    if (this.indicator !== indicator) {
        // disconnect old handlers
       this.indicator = indicator;
       // reconnect handlers
    }
} 
```

Then you could set the source indicator and update the tooltip without animation.

But it's much easier to just have one tooltip per indicator.

**Comment by JasonLG1979** [comment]
@andyholmes the brains of the operation is really the Constraint and as far as I can tell it works pretty well. But if you have any ideas on making it work better or be more GObject-y or in general more "correct" I'm open for anything you have to add. Again, I'm self taught. I have zero formal training or work experience when it comes to coding. My code I'm sure violates convention.

**Comment by JasonLG1979** [comment]
@andyholmes I updated the toolTip. I decided against being able to pass the tooltip around. Generally I think that it's more complicated then it's worth. It makes more sense to just have a separate tooltip for each indicator. I also decided against browse mode because to me a show timeout delay of 60ms presents an accessibility issue. In the event of rapid updates or fast mouse overs it can potentially cause a strobing effect. The last thing I want to do is cause someone to have a seizure. I don't hate users so much that I want to kill any of them,lol!!!
